File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res=[] #list of list
        part=[]
        def dfs(i):
            if i>=len(s):
                res.append(part.copy())
                return
            for j in range(i, len(s)):
                if self.isPalindrome(s,i,j):
                    part.append(s[i:j+1])
                    dfs(j+1)
                    part.pop()
        dfs(0)
        return res
    def isPalindrome(self,s,start,end):
        while start<end:
            if s[start]!=s[end]:
                return False
            start+=1
            end-=1
        return True
            
        

File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        if(s.empty()) return ans;
        vector<string> path;
        dfs(0,s,path,ans);
        return ans;
    }
    void dfs(int idx,string &s,vector<string>&path,vector<vector<string>>&ans){
        if(idx==s.size()){
            ans.push_back(path);
            return;
        }
        for(int j=idx;j<s.size();j++){
            if(ispalind(s,idx,j)){
                path.push_back(s.substr(idx,j-idx+1));
                dfs(j+1,s,path,ans);
                path.pop_back();
            }
        }
    }

    bool ispalind(string &s,int i,int j){
        while(i<=j){
            if(s[i++]!=s[j--]){
                return false;
            }
        }
        return true;
    }
};